realm/realm-library/src/main/java/io/realm/Realm.java

```diff
@@ -1157,6 +1160,160 @@ public void run() {
+                final Realm bgRealm = Realm.getInstance(realmConfiguration);
+                bgRealm.beginTransaction();
+                try {
+                    transaction.execute(bgRealm);
+
+                    if (!Thread.currentThread().isInterrupted()) {
+                        bgRealm.commitTransaction(new Runnable() {
+                            @Override
+                            public void run() {
+                                // The bgRealm needs to be closed before post event to caller's handler to avoid
+                                // concurrency problem. eg.: User wants to delete Realm in the callbacks.
+                                // This will close Realm before sending REALM_CHANGED.
+                                bgRealm.close();
+                            }
+                        });
+                        transactionCommitted = true;
+                    }
+                } catch (final Throwable e) {
+                    exception[0] = e;
+                } finally {
+                    if (!bgRealm.isClosed()) {
+                        if (bgRealm.isInTransaction()) {
+                            bgRealm.cancelTransaction();
+                        } else if (exception[0] != null) {
+                            RealmLog.w("Could not cancel transaction, not currently in a transaction.");
+                        }
+                        bgRealm.close();
+                    }
+
+                    final Throwable backgroundException = exception[0];
+                    // Send response as the final step to ensure the bg thread quit before others get the response!
+                    if (handler != null
+                            && !Thread.currentThread().isInterrupted()
+                            && handler.getLooper().getThread().isAlive()) {
+                        if (onSuccess != null && transactionCommitted) {
+                            handler.post(new Runnable() {
+                                @Override
+                                public void run() {
+                                    onSuccess.onSuccess();
+                                }
+                            });
+                        }
+
+                        if (backgroundException != null) {
+                            if (onError != null) {
+                                handler.post(new Runnable() {
+                                    @Override
+                                    public void run() {
+                                        onError.onError(backgroundException);
+                                    }
+                                });
+                            } else {
+                                handler.post(new Runnable() {
+                                    @Override
+                                    public void run() {
+                                        throw new RuntimeException(backgroundException);
```

Comment:
We shouldn't wrap exceptions in `RuntimeException` you might convert `Error` to a `RuntimeException`?